### ISSUE TITLE
Weird SubsetRandomSampler behavior

### ISSUE THREAD

**Issue by ishmeals** [pytorch/pytorch#137852]
### 🐛 Describe the bug

My model struggles to train when using `SubsetRandomSampler` (accuracy starts around 10% epoch 1). Implementing the same functionality with different code fixes it (accuracy 60% epoch 1).  

To me,
```python
train_dataloader = DataLoader(train_dataset, sampler=SubsetRandomSampler(train_mask))
val_dataloader = DataLoader(train_dataset, sampler=SubsetRandomSampler(val_mask))
```

is functionally the same as 

```python
train_dataset1, val_dataset1 = Subset(train_dataset, train_mask), Subset(train_dataset, val_mask)
train_dataloader = DataLoader(train_dataset1, shuffle=True)
val_dataloader = DataLoader(val_dataset1, shuffle=True)
```

If there is a difference could you let me know? Otherwise I don't see why swapping these two should consistently give such drastically different results.

Here is a full example. I was not able to narrow the diagnosis. Looking at the data returned by `SubsetRandomSampler` shows no obvious issues. I tested the below code both on `cpu` and `cuda` in Google Colab.

```python
import torch
from torch import nn
from torch.utils.data import DataLoader
from torchvision import datasets
from torchvision.transforms import ToTensor

train_dataset = datasets.FashionMNIST(
    root="data",
    train=True,
    download=True,
    transform=ToTensor(),
)

from torch.utils.data import Subset, SubsetRandomSampler
train_mask = range(50000)
val_mask = range(50000, 60000)

train_dataloader = DataLoader(train_dataset, sampler=SubsetRandomSampler(train_mask))
val_dataloader = DataLoader(train_dataset, sampler=SubsetRandomSampler(val_mask))

# Swap the above with this to drastically improve performance
# train_dataset1, val_dataset1 = Subset(train_dataset, train_mask), Subset(train_dataset, val_mask)
# train_dataloader = DataLoader(train_dataset1, shuffle=True)
# val_dataloader = DataLoader(val_dataset1, shuffle=True)

class NeuralNetwork(nn.Module):
    def __init__(self):
        super().__init__()
        self.flatten = nn.Flatten()
        self.linear_relu_stack = nn.Sequential(
            nn.Linear(28*28, 512),
            nn.ReLU(),
            nn.Linear(512, 512),
            nn.ReLU(),
            nn.Linear(512, 10)
        )

    def forward(self, x):
        x = self.flatten(x)
        logits = self.linear_relu_stack(x)
        return logits
model = NeuralNetwork()

loss_fn = nn.CrossEntropyLoss()
optimizer = torch.optim.SGD(model.parameters(), lr=1e-3)

def train(dataloader, model, loss_fn, optimizer):
    size = len(dataloader.dataset)
    model.train()
    for batch, (X, y) in enumerate(dataloader):
        # Compute prediction error
        pred = model(X)
        loss = loss_fn(pred, y)

        # Backpropagation
        loss.backward()
        optimizer.step()
        optimizer.zero_grad()

        if batch % 100 == 0:
            loss, current = loss.item(), (batch + 1) * len(X)
            print(f"loss: {loss:>7f}  [{current:>5d}/{size:>5d}]")

def test(dataloader, model, loss_fn):
    size = len(dataloader.dataset)
    num_batches = len(dataloader)
    model.eval()
    test_loss, correct = 0, 0
    with torch.no_grad():
        for X, y in dataloader:
            pred = model(X)
            test_loss += loss_fn(pred, y).item()
            correct += (pred.argmax(1) == y).type(torch.float).sum().item()
    test_loss /= num_batches
    correct /= size
    print(f"Test Error: \n Accuracy: {(100*correct):>0.1f}%, Avg loss: {test_loss:>8f} \n")

train(train_dataloader, model, loss_fn, optimizer)
test(val_dataloader, model, loss_fn)
```

### Versions

Collecting environment information...
PyTorch version: 2.4.1+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.30.4
Libc version: glibc-2.35

Python version: 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.1.85+-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.2.140
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 535.104.05
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.6
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               2
On-line CPU(s) list:                  0,1
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) CPU @ 2.00GHz
CPU family:                           6
Model:                                85
Thread(s) per core:                   2
Core(s) per socket:                   1
Socket(s):                            1
Stepping:                             3
BogoMIPS:                             4000.30
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat md_clear arch_capabilities
Hypervisor vendor:                    KVM
Virtualization type:                  full
L1d cache:                            32 KiB (1 instance)
L1i cache:                            32 KiB (1 instance)
L2 cache:                             1 MiB (1 instance)
L3 cache:                             38.5 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0,1
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Mitigation; PTE Inversion
Vulnerability Mds:                    Vulnerable; SMT Host state unknown
Vulnerability Meltdown:               Vulnerable
Vulnerability Mmio stale data:        Vulnerable
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Vulnerable
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Vulnerable
Vulnerability Spectre v1:             Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:             Vulnerable; IBPB: disabled; STIBP: disabled; PBRSB-eIBRS: Not affected; BHI: Vulnerable (Syscall hardening enabled)
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Vulnerable

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] optree==0.13.0
[pip3] torch==2.4.1+cu121
[pip3] torchaudio==2.4.1+cu121
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.19.1+cu121
[conda] Could not collect

cc @andrewkho @divyanshk @SsnL @VitalyFedyunin @dzhulgakov

**Comment by sumanthdonthula** [comment]
Hi @ishmeals,

I assume that you are getting two different accuracies because random sampling select items in a dataset randomly with no order for example [[1,2],[1,3],[1,4]] if you want one sample of 2 data points from indexes [0,1,2] with random sampling you get [[1,2],[1,3]] or [[1,2],[1,4]] or [[1,3],[1,4]] with no order.

Where as with Subset you get ordered samples like [[1,2],[1,3]] order is maintained.

Talking about accuracy, yes as you use subset the data points are structured and as you have more samples of data for particular digit, this will align with validations as you might have same kind of digit for which model have already seen earlier. Where as in random sampling case it has not seen enough samples of image. This will get better with epochs.

Replicate this code to understand the two methods on a sample dataset:

import torch
from torch.utils.data import DataLoader, SubsetRandomSampler, TensorDataset

# Create a simple dataset
data = torch.randn(10, 2)  # 10 samples, each with 2 features
targets = torch.randint(0, 2, (10,))  # Binary targets

dataset = TensorDataset(data, targets)

for i in range(len(dataset)):
    data_point, target = dataset[i]
    print(f"Data Point {i}: {data_point.numpy()}, Target: {target.item()}")

# Define the indices you want to sample from
indices = [0, 1, 2, 3]

print("----------SubsetRandomSampler-------------")
# Create a SubsetRandomSampler
sampler = SubsetRandomSampler(indices)

# Create a DataLoader using the sampler
dataloader = DataLoader(dataset, sampler=sampler, batch_size=2)

# Iterate through the DataLoader
for batch in dataloader:
    print(batch)

print("----------Subset-------------")
from torch.utils.data import Subset

# Create a Subset from the original dataset using specific indices

subset = Subset(dataset, indices)

# Create a DataLoader for the subset
subset_loader = DataLoader(subset, batch_size=2)

# Iterate through the DataLoader for the subset
for batch in subset_loader:
    print(batch)

**Comment by ishmeals** [comment]
Hello,

Thanks for the response. I include `shuffle=True`, so I believe in both situations data is getting shuffled. Please let me know if I misunderstood.

**Comment by sumanthdonthula** [comment]
I think with "shuffle" involved. I think the only difference I can possibly think of is, 

subset random sampler selects data points from dataset without replacement. 

Where as in second case it shuffles dataset and selects the, say for example it shuffles 50k data points from data set and outputs 64 sample from specific positions of the overall dataset. I sense there is possibility of replacement as shuffling does not involve removing used samples.

I still wanted to listen from others too! Let's see if my interpretation is correct or not.

**Comment by ishmeals** [comment]
I believe in the second example, there is no replacement either.
